packages/x-data-grid/src/hooks/features/pagination/useGridRowCount.ts

```diff
@@ -109,6 +119,49 @@ export const useGridRowCount = (
   useGridRegisterPipeProcessor(apiRef, 'exportState', stateExportPreProcessing);
   useGridRegisterPipeProcessor(apiRef, 'restoreState', stateRestorePreProcessing);
 
+  /**
+   * EVENTS
+   */
+  const handlePaginationMetaChange = React.useCallback(
+    (meta: GridPaginationState['meta']) => {
+      if (props.paginationMode === 'client') {
+        return;
+      }
+
+      if (!meta.hasNextPage && rowCountState === -1) {
+        apiRef.current.setRowCount(
+          paginationModel.pageSize * paginationModel.page + paginationModel.pageSize,
```

Comment:
```suggestion
          paginationModel.pageSize * paginationModel.page + gridRowCountSelector(apiRef),
```

How about doing this instead?
Then users won't have to set rowCount on the last page: https://github.com/mui/mui-x/pull/12474/files#diff-d5340f1a8a49834c73bed351907ef06eceda3faba1ca045fc35da9ca17a41932R51-R56

Reply:
Oh, I was too late with this suggestion 😅

Reply:
Unfortunately, this may not work in some scenarios since the handler could run before (when the meta updates), while the rows would update later, also we need to consider the top-level rows here, to be able to work fine with nested data.

I removed this handler and moved the logic to the `useEffect` to react to the rows update better.

Let me know how you feel about it.

Reply:
Looks great!